interp/src/interpreter/interpret_group.rs

```diff
@@ -1,357 +1,479 @@
+                    let v = std::mem::take(mut_ref);
 
-                        // update the cell state in write_env
-                        write_env = primitives::update_cell_state(
-                            &dst_cell,
-                            &inputs,
-                            &outputs,
-                            &write_env,
-                            component.clone(),
-                        )?;
+                    *mut_ref = if v.is_tlv() {
+                        v.unwrap_tlv().unlock().into()
+                    } else {
+                        // this branch should be impossible since the list of
+                        // ports we're iterating over are only those w/ updates
+                        unreachable!()
                     }
-                } else {
-                    // otherwise, add the write to the update queue; currently only handles registers
-
-                    // get input cell
-                    let inputs = vec![src_cell.clone()];
-                    // get dst_cell's output port
-                    let outputs = vec![assign.dst.borrow().name.clone()];
-
-                    uq = uq.init_cells(
-                        &dst_cell,
-                        inputs,
-                        outputs,
-                        write_env.clone(),
-                    );
                 }
             }
+        } else if !non_comb_cells.is_empty() {
+            let tmp = std::mem::take(&mut non_comb_cells);
+
+            let new_assigns = eval_prims(
+                &mut working_env,
+                &dependency_map,
+                tmp.iter().map(|x| x.into()),
+                false,
+            );
+            assign_worklist.extend(new_assigns.into_iter())
+        }
+        // all are empty
+        else {
+            let assigns: WorkList = working_env
+                .do_tick()
+                .into_iter()
+                .filter_map(|port| dependency_map.map.get(&port))
+                .flatten()
+                .cloned()
+                .collect();
+
+            assign_worklist = assigns;
         }
-        write_env = uq.do_tick(write_env.clone())?;
-        counter += 1;
     }
-    Ok(write_env)
+
+    Ok(working_env.collapse_env())
 }
 
-/// Evaluate guard implementation
-#[allow(clippy::borrowed_box)]
-// XXX: Allow for this warning. It would make sense to use a reference when we
-// have the `box` match pattern available in Rust.
-fn eval_guard(comp: &ir::Id, guard: &Box<ir::Guard>, env: &Environment) -> u64 {
-    (match &**guard {
-        ir::Guard::Or(g1, g2) => {
-            (eval_guard(comp, g1, env) == 1) || (eval_guard(comp, g2, env) == 1)
-        }
-        ir::Guard::And(g1, g2) => {
-            (eval_guard(comp, g1, env) == 1) && (eval_guard(comp, g2, env) == 1)
+// yes I know we don't need two different lifetimes here but they are
+// meaningfully separate so fight me
+fn eval_prims<'a, 'b, I: Iterator<Item = &'b RRC<ir::Cell>>>(
```

Comment:
Better comment explaining why the lifetimes are different.

Reply:
Fiiiinnneeeeeeeeeeeeeee. I take personal offense to you not humoring me